### PR TITLE
Add lora device customization: downstream and preferred receive window

### PR DESCRIPTION
**Pull request by fbeltrao**
Changes:
- new twin property to set device preferred receive window
- adjust C2D message timeout based on device preferred receive window
- use fix timeout to check for additional C2D message (20ms)
- new twin property to enable/disable downstream messages
- add channels 3-6 in CI test device
- use default direct method handler to handle cache reset
- Add documentation for lora device customisation
- fix RX2 dr and frequency in CI

Fixes: AB#654, AB#664, AB#922

### CODE REVIEW

**Comment by fbeltrao** [comment]
Merged from dev due to conflicts, waiting for CI to succeed

**Comment by fbeltrao** [comment]
I will have to review the unit tests added to validate c2d timing. Fails in CI but works locally.

File: LoRaEngine/test/LoRaWan.IntegrationTest/C2DMessageTest.cs [hunk 1]
```diff
@@ -563,5 +559,200 @@ public async Task C2D_Using_Invalid_FPort_Should_Be_Ignored(byte fport, string d
...
+            if (!foundReceivePacketInRX2)
+            {
+                foundReceivePacketInRX2 = this.ArduinoDevice.SerialLogs.Any(x => x.StartsWith(expectedRxSerial2));
+            }
+
+            Assert.True(foundReceivePacketInRX2, $"Could not find lora receiving message '{expectedRxSerial2}'");
+        }
+
+        // [Fact]
+        // public async Task C2D_With_PreferredWindow2_Should_Process_C2d_In_First_Upstream()
```
Comment (Mandur) [inline LoRaEngine/test/LoRaWan.IntegrationTest/C2DMessageTest.cs]: Is this test needed for later stage or is this just a legacy item?

File: LoRaEngine/test/LoRaWan.IntegrationTest/LoRaArduinoSerial.Helper.cs [hunk 2]
```diff
@@ -74,8 +74,12 @@ public async Task SetupLora(LoraRegion region)
                     await this.setChannelAsync(0, 868.1F);
                     await this.setChannelAsync(1, 868.3F);
                     await this.setChannelAsync(2, 868.5F);
+                    await this.setChannelAsync(3, 867.1F);
```
Comment (Mandur) [inline LoRaEngine/test/LoRaWan.IntegrationTest/LoRaArduinoSerial.Helper.cs]: Is there a reason you chose this additional channels?

File: LoRaEngine/test/LoRaWanNetworkServer.Test/MessageProcessorSingleGatewayTest.cs [hunk 3]
```diff
@@ -260,87 +260,6 @@ public async Task OTAA_Unconfirmed_Message_With_FcntUp_10_Should_Send_Data_To_Io
             Assert.False(loraDevice.HasFrameCountChanges);
         }
 
-        [Fact]
-        public async Task OTAA_Unconfirmed_With_Cloud_To_Device_Message_Returns_Downstream_Message()
```
Comment (Mandur) [inline LoRaEngine/test/LoRaWanNetworkServer.Test/MessageProcessorSingleGatewayTest.cs]: Is the deletion of this test intentional?

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/MessageProcessor.cs [hunk 4]
```diff
@@ -247,10 +243,11 @@ async Task<DownlinkPktFwdMessage> ProcessLoRaMessageAsync(LoRaTools.LoRaPhysical
                                 }
                             }
 
-                            // What do we need to send an UpAck to IoT Hub?
-                            // What is the content of the message
-                            // TODO Future: Don't wait if it is an unconfirmed message
-                            await this.SendDeviceEventAsync(loRaDevice, rxpk, payloadData, loraPayload, timeWatcher);
+                            if (!await this.SendDeviceEventAsync(loRaDevice, rxpk, payloadData, loraPayload, timeWatcher))
+                            {
+                                // failed to send event to IoT Hub, stop now
```
[the remark this answers is not in the document]
Comment (fbeltrao) [reply LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/MessageProcessor.cs]: This is being logged here:
https://github.com/Azure/iotedge-lorawan-starterkit/blob/fbe-dev-disablec2d-always2ndwindow/LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs#L151

File: LoRaEngine/test/LoRaWan.IntegrationTest/LoRaArduinoSerial.Helper.cs (hunk 2 above)
Comment (fbeltrao) [reply LoRaEngine/test/LoRaWan.IntegrationTest/LoRaArduinoSerial.Helper.cs]: They were part of the Arduino code, we forgot to add when we started

File: README.md [hunk 5]
```diff
@@ -239,6 +236,17 @@ It should look something like this for ABP:
 - Click Save
 - Turn on the device and you are ready to go
 
+### Optional device properties
+
+Customizations to lora devices are set by creating specific twin desired properties on the device. The following customizations are available:
+
+|Name|Description|Configuration|When to use|
+|-|-|-|-|
+|Enable/disable downstream messages|Allows disabling the downstream (cloud to device) for a device. By default downstream messages are enabled| Add twin desired property `"Downlink": false` to disable downstream messages. The absence of the twin property or setting value to `true` will enable downlink messages.|Disabling downlink on devices decreases message processing latency, since the network server will not look for cloud to device messages when an uplink is received. Only disable it in devices that are not expecting messages from cloud|
```
Comment (ronniesa) [inline README.md]: acknowledgement of confirmed upstream are sent to devices even when downlink is set to false

File: LoRaEngine/test/LoRaWanNetworkServer.Test/MessageProcessorSingleGatewayTest.cs (hunk 3 above)
Comment (fbeltrao) [reply LoRaEngine/test/LoRaWanNetworkServer.Test/MessageProcessorSingleGatewayTest.cs]: Moved to here https://github.com/Azure/iotedge-lorawan-starterkit/blob/fbe-dev-disablec2d-always2ndwindow/LoRaEngine/test/LoRaWanNetworkServer.Test/MessageProcessor_End2End_NoDep_CloudToDeviceMessage_Tests.cs

File: LoRaEngine/test/LoRaWan.IntegrationTest/C2DMessageTest.cs (hunk 1 above)
Comment (fbeltrao) [reply LoRaEngine/test/LoRaWan.IntegrationTest/C2DMessageTest.cs]: It is the test I used to build the feature. I will remove it.

File: README.md [hunk 6]
```diff
@@ -239,6 +236,17 @@ It should look something like this for ABP:
...
+|Preferred receive window|Allows setting the device preferred receive window (RX1 or RX2). The default preferred receive window is 1| Add twin desired property `"PreferredWindow": 2` sets RX2 as preferred window. The absence of the twin property or setting the value to `1` will set RX1 as preferred window.|Using RX2 as preferred window will cause the network server to allocate more time when checking for cloud to device messages. The increased allocated time improves chances that cloud to device messages will arrive the lora device with fewer uplinks. The downside is that the device will only receive downlink messages in RX2. Recommended in scenarios that a cloud to device message needs to be sent as a response to an upstream message (i.e. setting current time)|
+
```
Comment (ronniesa) [inline README.md]: Using second window increase the chances that the end application can process the upstream message and send a cloud to device message to the lora device without requiring and additional upstream message. Basically completing the round trip in less than 2 seconds.

**Pull request merged**